bliss/models/fnp.py

```diff
@@ -0,0 +1,475 @@
+
+    This implementation is based on https://github.com/AMLab-Amsterdam/FNP,
+    distributed with the MIT license. Significant changes have been made,
+    so any problems or bugs should not be attributed to the original authors.
+    """
+
+    def __init__(self, dim_u, temperature=0.3):
+        super().__init__()
+        ## Temperature for LogitRelaxedBernoulli when training
+        self.temperature = temperature
+        ## Initialized the distance function pairwise_g
+        ## This has a single learned parameter
+        self.g_logscale = nn.Parameter(torch.tensor(math.log(dim_u) * 0.5))
+        self.g = lambda x: self.logitexp(
+            -0.5
+            * torch.sum(torch.pow(x[:, dim_u:] - x[:, 0:dim_u], 2), 1, keepdim=True)
+            / self.g_logscale.exp()
+        ).view(x.size(0), 1)
+
+    def sample_G(self, uR):
+        # get the indices of an upper triangular adjacency matrix that represents the DAG
+        idx_utr = np.triu_indices(uR.size(0), 1)
+
+        # get the ordering
+        ordering = self.order_z(uR)
+        # sort the latents according to the ordering
+        sort_idx = torch.sort(torch.squeeze(ordering), 0)[1]
+        Y = uR[sort_idx, :]
+        # form the latent pairs for the edges
+        Z_pairs = torch.cat([Y[idx_utr[0]], Y[idx_utr[1]]], 1)
+        # get the logits for the edges in the DAG
+        logits = self.g(Z_pairs)
+
+        if self.training:
+            p_edges = LogitRelaxedBernoulli(temperature=self.temperature, logits=logits)
+            G = torch.sigmoid(p_edges.rsample())
+        else:
+            p_edges = Bernoulli(logits=logits)
+            G = p_edges.sample()
+
+        # embed the upper triangular to the adjacency matrix
+        unsorted_G = torch.zeros(uR.size(0), uR.size(0), device=uR.device)
+        unsorted_G[idx_utr[0], idx_utr[1]] = G.squeeze()
+        # unsort the dag to conform to the data order
+        original_idx = torch.sort(sort_idx)[1]
+        unsorted_G = unsorted_G[original_idx, :][:, original_idx]
+
+        return unsorted_G
+
+    def sample_A(self, uM, uR):
+        indices = []
+        for element in product(range(uM.size(0)), range(uR.size(0))):
+            indices.append(element)
+        indices = np.array(indices)
+        Z_pairs = torch.cat([uM[indices[:, 0]], uR[indices[:, 1]]], 1)
+
+        logits = self.g(Z_pairs)
+        if self.training:
+            p_edges = LogitRelaxedBernoulli(temperature=self.temperature, logits=logits)
+            A_vals = torch.sigmoid(p_edges.rsample())
+        else:
+            p_edges = Bernoulli(logits=logits)
+            A_vals = p_edges.sample()
+
+        # embed the values to the adjacency matrix
+        A = torch.zeros(uM.size(0), uR.size(0), device=uM.device)
+        A[indices[:, 0], indices[:, 1]] = A_vals.squeeze()
+
+        return A
+
+    @staticmethod
+    def logitexp(logp):
+        # https: // github.com / pytorch / pytorch / issues / 4007
```

Comment:
The issue you are pointing too seems to have been merged and resolved. Do you still need this separate function?

Reply:
I'm not really sure why the FNP authors linked to that issue, so I removed that comment and replaced log with log1p where appropriate.

I don't think there is a built-in to torch that does this function